### PR TITLE
Implement a Convex Hull method

### PR DESCRIPTION
**Pull request by urschrei**
The current implementation is slow (the worst case is On<sup>4</sup>) compared to e.g.
[Quickhull](http://www.cse.yorku.ca/~aaw/Hang/quick_hull/Algorithm.html) (I struggled with the recursion + retained points implementation), but it's a reasonable starting point, and a convex hull method is the fundamental building block for Delaunay triangulation and Voronoi tessellation methods.

### CODE REVIEW

File: src/algorithm/convexhull.rs [hunk 1]
```diff
@@ -0,0 +1,174 @@
...
+        let u = (dot11 * dot02 - dot01 * dot12) * inv_denom;
+        let v = (dot00 * dot12 - dot01 * dot02) * inv_denom;
+        (u > T::zero()) && (v > T::zero()) && (u + v < T::one())
+    }
+}
+
+// Adapted from http://codereview.stackexchange.com/a/141752/2630
+// The algorithm is from Heineman, G.T., Pollice, G., Selkow, S., 2008.
+// "Algorithms in a Nutshell". O’Reilly Media, Inc., pp261–8
+pub fn convex_hull<T>(points: &BTreeSet<Point<T>>) -> Vec<Point<T>>
```
Comment (frewsxcv) [inline src/algorithm/convexhull.rs]: This doesn't need to be `pub`, right?

File: src/algorithm/convexhull.rs [hunk 2]
```diff
@@ -0,0 +1,174 @@
+use std::cmp::Ordering;
+use num_traits::Float;
+use types::{Point, Polygon, LineString};
+use std::collections::BTreeSet;
+
+impl<T> Eq for Point<T> where T: Float {}
```
Comment (frewsxcv) [inline src/algorithm/convexhull.rs]: Why is this needed?

File: src/algorithm/convexhull.rs [hunk 3]
```diff
@@ -0,0 +1,174 @@
...
+    // when that's equal, sort by distance to head
+    hull.sort_by(|a, b| {
+        let angle_a = head.angle(&a);
+        let angle_b = head.angle(&b);
+        angle_a.partial_cmp(&angle_b).unwrap()
+    });
+    // we need to close the Polygon
+    let final_element = hull[0].clone();
+    hull.push(final_element);
+    hull.into_iter().collect::<Vec<Point<T>>>()
```
Comment (frewsxcv) [inline src/algorithm/convexhull.rs]: Prior to this line, is `hull` not already a `Vec<Point<T>>`?

File: src/algorithm/convexhull.rs [hunk 4]
```diff
@@ -0,0 +1,174 @@
...
+    /// let res = poly.convexhull();
+    /// assert_eq!(res.exterior, correct_hull);
+    /// ```
+    fn convexhull(&self) -> Self where T: Float;
+}
+
+impl<T> Convexhull<T> for Polygon<T>
+    where T: Float
+{
+    fn convexhull(&self) -> Polygon<T> {
```
Comment (frewsxcv) [inline src/algorithm/convexhull.rs]: Nit: If "convex hull" is two words, I'd personally call this `ConvexHull::convex_hull`

File: src/algorithm/convexhull.rs (hunk 2 above)
Comment (urschrei) [inline src/algorithm/convexhull.rs]: `Ord` needs `Eq`, and deriving it in `types.rs` didn't work…

File: src/algorithm/convexhull.rs (hunk 1 above)
Comment (urschrei) [inline src/algorithm/convexhull.rs]: Whoops, nope.

File: src/algorithm/convexhull.rs (hunk 4 above)
Comment (urschrei) [inline src/algorithm/convexhull.rs]: Fixed!

File: src/algorithm/convexhull.rs (hunk 2 above)
Comment (frewsxcv) [inline src/algorithm/convexhull.rs]: Is there some reason you need `Ord`? IIRC, `num::Float` intentionally doesn't implement `Ord` since it can't guarantee total order...I think?

Comment (urschrei) [inline src/algorithm/convexhull.rs]: This is a problem related to punting on `NaN` values – BTreeSet needs `Ord`, but as you say, there's no total ordering for types which included `NaN`. However, I'm not sure that it should be possible to construct geometries with NaN values in the first place.
I feel like there should be a more elegant solution here, but it'll need input from someone who knows more than I do about floating point…

Comment (frewsxcv) [inline src/algorithm/convexhull.rs]: > `BTreeSet` needs `Ord`

Ah right, forgot about this. What you have is fine for now. Can you just add a comment next to the `Ord` implementation stating that it's needed for `BTreeSet`?

Comment (urschrei) [inline src/algorithm/convexhull.rs]: Done!
